cli/src/main/java/org/eclipse/hono/cli/application/ApplicationProfile.java

```diff
@@ -0,0 +1,147 @@
+import java.util.stream.Collectors;
+
+import org.eclipse.hono.cli.AbstractCliClient;
+import org.eclipse.hono.client.ApplicationClientFactory;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.config.ClientConfigProperties;
+import org.springframework.core.MethodParameter;
+import org.springframework.shell.CompletionContext;
+import org.springframework.shell.CompletionProposal;
+import org.springframework.shell.standard.ShellCommandGroup;
+import org.springframework.shell.standard.ShellComponent;
+import org.springframework.shell.standard.ShellMethod;
+import org.springframework.shell.standard.ShellMethodAvailability;
+import org.springframework.shell.standard.ShellOption;
+import org.springframework.shell.standard.ValueProviderSupport;
+import org.springframework.stereotype.Component;
+
+/**
+ * This class contains the methods that the user can execute acting as an application consumer.
+ * To execute these commands the ClientConfig.profile must be valued as APPLICATION and the connection must be created
+ * with the sub command connect of the shell.
+ */
+@ShellComponent
+@ShellCommandGroup("Northbound Commands (Application)")
+public class ApplicationProfile extends AbstractCliClient {
+    /**
+     * Message types.
+     */
+    public static final String TYPE_TELEMETRY = "telemetry";
+    public static final String TYPE_EVENT = "event";
+    public static final String TYPE_ALL = "all";
+
+    /**
+     * To stop the executions of internal commands.
+     */
+    private CountDownLatch connLatch;
+
+    /*
+     * This command allow the creation of the ApplicationClientFactory.
+     */
+    @ShellMethod("Create the connection to a given endpoint\n" +
+            "\t\t\t[-H --host] [-P --port] [-u --username] [-p --password]")
+    void connectApplication(@ShellOption(value = {"-H", "--host"},  defaultValue = "", help = "The address of the endpoint which you want to establish a connection to") final String host,
+                 @ShellOption(value = {"-P", "--port"}, defaultValue = "0", help = "The port used by the service on the connection endpoint") final int port,
+                 @ShellOption(value = {"-u", "--username"}, defaultValue = "", help = "Credentials to use to connect to the endpoint") final String username,
+                 @ShellOption(value = {"-p", "--password"}, defaultValue = "", help = "Credentials to use to connect to the endpoint") final String password) {
+        if (clientFactory != null) {
+            shellHelper.printInfo("The previously connection will be destroyed");
+        }
+        final ClientConfigProperties cfp = new ClientConfigProperties(this.honoClientConfig);
+        cfp.setHost(host);
+        cfp.setPort(port);
+        cfp.setUsername(username);
+        cfp.setPassword(password);
+        //create the factory with connection info
+        try {
+            clientFactory = ApplicationClientFactory.create(HonoConnection.newConnection(vertx, cfp));
```

Comment:
this should not only create the connection factory but should actually establish the connection. The connection can then be held in a field and can be used to determine the availability state for the commands (*HonoConnection.isConnected()*). FMPOV there is no reason for connecting/disconnecting for every command as there also is a certain overhead involved ... Plus, the methods start actually doing what their names suggest ;-)

Reply:
FMPOV is the best approach since for slow/instable connection you can stop the execution of commands (without closing the client) still having stale messages in/outgoing that can make noise in the next execution. So the overhead is justified

Reply:
The execution of a command can be stopped anyway, regardless of whether the connection is already established or is being established as part of the command.
*Execution of a command* means sending an AMQP frame and (often) receiving an AMQP frame in response. I can always stop waiting for the response to arrive. There is no need to establish a new connection for every command ...